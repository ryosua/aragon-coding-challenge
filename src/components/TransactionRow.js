import React from 'react'
import PropTypes from 'prop-types'
import { TableRow, TableCell, Text } from '@aragon/ui'

const TransactionRow = ({ hash, from, to, value }) => (
    <TableRow>
        <TableCell>
            <Text>{hash}</Text>
        </TableCell>
        <TableCell>
            <Text>{from.address}</Text>
        </TableCell>
        <TableCell>
            <Text>{to.address}</Text>
        </TableCell>
        <TableCell>
            <Text>{value}</Text>
        </TableCell>
    </TableRow>
)

TransactionRow.propTypes = {
    hash: PropTypes.string.isRequired,
    from: PropTypes.object.isRequired,
    to: PropTypes.object.isRequired,
    value: PropTypes.number.isRequired
}

export default TransactionRow
