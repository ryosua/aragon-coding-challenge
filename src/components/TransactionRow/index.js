import React from 'react'
import PropTypes from 'prop-types'
import { TableRow, TableCell, Text } from '@aragon/ui'
import { convertWeiToUSD } from 'api/getConversionRate'
import RateContext from 'contexts/RateContext'

const TransactionRow = ({ hash, from, to, value }) => {
    return (
        <RateContext.Consumer>
            {({ ethToUSDRate }) => {
                const valueInUSD = ethToUSDRate ? convertWeiToUSD(value, ethToUSDRate) : 'loading'
                return (
                    <TableRow>
                        <TableCell>
                            <Text>{hash}</Text>
                        </TableCell>
                        <TableCell>
                            <Text>{from.address}</Text>
                        </TableCell>
                        <TableCell>
                            <Text>{to.address}</Text>
                        </TableCell>
                        <TableCell>
                            <Text>{valueInUSD}</Text>
                        </TableCell>
                    </TableRow>
                )
            }}
        </RateContext.Consumer>
    )
}

TransactionRow.propTypes = {
    hash: PropTypes.string.isRequired,
    from: PropTypes.object.isRequired,
    to: PropTypes.object.isRequired,
    value: PropTypes.number.isRequired
}

export default TransactionRow
