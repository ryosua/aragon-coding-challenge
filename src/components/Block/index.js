import React from 'react'
import PropTypes from 'prop-types'
import map from 'lodash/map'
import isEmpty from 'lodash/isEmpty'
import text from 'constants/text'
import { Table, TableHeader, TableRow, Text } from '@aragon/ui'
import TransactionRow from 'components/TransactionRow'

const { blockNumber: blockNumberText, hash, from, to, value } = text.block

const Block = ({ blockNumber, transactions, emptyBlockText }) => (
    <>
        <Text size={'large'}>{`${blockNumberText} ${blockNumber}`}</Text> <br />
        {isEmpty(transactions) ? (
            <Text className="empty-text">{emptyBlockText}</Text>
        ) : (
            <Table
                header={
                    <TableRow>
                        <TableHeader title={hash} />
                        <TableHeader title={from} />
                        <TableHeader title={to} />
                        <TableHeader title={value} />
                    </TableRow>
                }>
                {map(transactions, transaction => (
                    <TransactionRow key={transaction.hash} {...transaction} />
                ))}
            </Table>
        )}
        <br />
    </>
)

Block.propTypes = {
    blockNumber: PropTypes.number.isRequired,
    transactions: PropTypes.arrayOf(
        PropTypes.shape({
            hash: PropTypes.string.isRequired,
            from: PropTypes.object.isRequired,
            to: PropTypes.object.isRequired,
            value: PropTypes.number.isRequired
        })
    ).isRequired,
    emptyBlockText: PropTypes.string.isRequired
}

export default Block
