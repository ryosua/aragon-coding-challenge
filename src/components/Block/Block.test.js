import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import Block from './index.js'
import TransactionRow from 'components/TransactionRow'

describe('Block tests', () => {
    it('Block renders according to snapshot.', () => {
        const block = {
            number: 6551120,
            transactions: [
                {
                    hash: '0xd67348d36f906ba7b2be09baa868e0e93cf310f8d06146ec8e0936e96410e05d',
                    nonce: 97438,
                    index: 1,
                    from: {
                        address: '0xBA826fEc90CEFdf6706858E5FbaFcb27A290Fbe0'
                    },
                    to: {
                        address: '0xD0a4b8946Cb52f0661273bfbC6fD0E0C75Fc6433'
                    },
                    value: 0,
                    gasPrice: 56450000000,
                    gas: 152609,
                    inputData:
                        '0xa9059cbb00000000000000000000000020ec03ea0bc655d10399f7dce2f9cd1e691b8ea9000000000000000000000000000000000000000000006d493e59700229ed3c00',
                    status: 'SUCCESS',
                    block: {
                        hash: '0x482ef4f5d0931339f8b939d2a49d8ae8a226d4e2e2cd9c442550b89f8b8726ec'
                    }
                },
                {
                    hash: '0x24c810eaa67a6ec5e84c75ad6656a1e1af14b6d1dcf69990079dc78e336563f7',
                    nonce: 103,
                    index: 7,
                    from: {
                        address: '0x16C598eE4540b73898dEd6684e78774928cE2e88'
                    },
                    to: {
                        address: '0x7C39a481E17CCAfbe8e50Ec1388fDBB01Bbb3bf4'
                    },
                    value: 0,
                    gasPrice: 29593750000,
                    gas: 60000,
                    inputData:
                        '0xa9059cbb0000000000000000000000008b6fb96247bd209ac27ba3ba86fbfac441c6515d0000000000000000000000000000000000000000000000056bc75e2d63100000',
                    status: 'SUCCESS',
                    block: {
                        hash: '0x482ef4f5d0931339f8b939d2a49d8ae8a226d4e2e2cd9c442550b89f8b8726ec'
                    }
                }
            ]
        }
        const tree = renderer.create(<Block blockNumber={block.number} transactions={block.transactions} />).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('Block renders an empty message if it contains no transactions.', () => {
        const block = {
            number: 6551120,
            transactions: []
        }

        const wrapper = shallow(<Block blockNumber={block.number} transactions={block.transactions} />)
        expect(wrapper.exists('.empty-text')).toEqual(true)
    })

    it('Block renders all transactions', () => {
        const block = {
            number: 6551120,
            transactions: [
                {
                    hash: '0xd67348d36f906ba7b2be09baa868e0e93cf310f8d06146ec8e0936e96410e05d',
                    nonce: 97438,
                    index: 1,
                    from: {
                        address: '0xBA826fEc90CEFdf6706858E5FbaFcb27A290Fbe0'
                    },
                    to: {
                        address: '0xD0a4b8946Cb52f0661273bfbC6fD0E0C75Fc6433'
                    },
                    value: 0,
                    gasPrice: 56450000000,
                    gas: 152609,
                    inputData:
                        '0xa9059cbb00000000000000000000000020ec03ea0bc655d10399f7dce2f9cd1e691b8ea9000000000000000000000000000000000000000000006d493e59700229ed3c00',
                    status: 'SUCCESS',
                    block: {
                        hash: '0x482ef4f5d0931339f8b939d2a49d8ae8a226d4e2e2cd9c442550b89f8b8726ec'
                    }
                },
                {
                    hash: '0x24c810eaa67a6ec5e84c75ad6656a1e1af14b6d1dcf69990079dc78e336563f7',
                    nonce: 103,
                    index: 7,
                    from: {
                        address: '0x16C598eE4540b73898dEd6684e78774928cE2e88'
                    },
                    to: {
                        address: '0x7C39a481E17CCAfbe8e50Ec1388fDBB01Bbb3bf4'
                    },
                    value: 0,
                    gasPrice: 29593750000,
                    gas: 60000,
                    inputData:
                        '0xa9059cbb0000000000000000000000008b6fb96247bd209ac27ba3ba86fbfac441c6515d0000000000000000000000000000000000000000000000056bc75e2d63100000',
                    status: 'SUCCESS',
                    block: {
                        hash: '0x482ef4f5d0931339f8b939d2a49d8ae8a226d4e2e2cd9c442550b89f8b8726ec'
                    }
                }
            ]
        }

        const wrapper = shallow(<Block blockNumber={block.number} transactions={block.transactions} />)
        expect(wrapper.find(TransactionRow)).toHaveLength(block.transactions.length)
    })
})
