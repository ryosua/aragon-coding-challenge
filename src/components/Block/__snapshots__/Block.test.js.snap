// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block tests Block renders according to snapshot. 1`] = `
Array [
  <span
    className="Text-yxw4o9-0 hzfDfW"
    size="large"
  >
    Block Number: 
    6551120
  </span>,
  " ",
  <br />,
  <table
    className="Table__StyledTable-uvcan9-0 ksKxdz"
  >
    <thead>
      <tr
        className="TableRow__StyledTableRow-sc-3jn6yz-0 cFtmyC"
      >
        <th
          className="TableHeader__StyledTableHeader-sc-1qxm8cp-0 hKOaZg"
        >
          <div
            className="Text-yxw4o9-0-div kRDWpm"
            color="#707070"
          >
            TxHash
          </div>
        </th>
        <th
          className="TableHeader__StyledTableHeader-sc-1qxm8cp-0 hKOaZg"
        >
          <div
            className="Text-yxw4o9-0-div kRDWpm"
            color="#707070"
          >
            From
          </div>
        </th>
        <th
          className="TableHeader__StyledTableHeader-sc-1qxm8cp-0 hKOaZg"
        >
          <div
            className="Text-yxw4o9-0-div kRDWpm"
            color="#707070"
          >
            To
          </div>
        </th>
        <th
          className="TableHeader__StyledTableHeader-sc-1qxm8cp-0 hKOaZg"
        >
          <div
            className="Text-yxw4o9-0-div kRDWpm"
            color="#707070"
          >
            Value
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="TableRow__StyledTableRow-sc-3jn6yz-0 cFtmyC"
      >
        <td
          className="TableCell__StyledTableCell-sc-110j155-0 bqABXJ"
        >
          <div
            className="TableCell__StyledTableCellContent-sc-110j155-1 klobR"
          >
            <span
              className="Text-yxw4o9-0 bIIgvx"
            >
              0xd67348d36f906ba7b2be09baa868e0e93cf310f8d06146ec8e0936e96410e05d
            </span>
          </div>
        </td>
        <td
          className="TableCell__StyledTableCell-sc-110j155-0 bqABXJ"
        >
          <div
            className="TableCell__StyledTableCellContent-sc-110j155-1 klobR"
          >
            <span
              className="Text-yxw4o9-0 bIIgvx"
            >
              0xBA826fEc90CEFdf6706858E5FbaFcb27A290Fbe0
            </span>
          </div>
        </td>
        <td
          className="TableCell__StyledTableCell-sc-110j155-0 bqABXJ"
        >
          <div
            className="TableCell__StyledTableCellContent-sc-110j155-1 klobR"
          >
            <span
              className="Text-yxw4o9-0 bIIgvx"
            >
              0xD0a4b8946Cb52f0661273bfbC6fD0E0C75Fc6433
            </span>
          </div>
        </td>
        <td
          className="TableCell__StyledTableCell-sc-110j155-0 bqABXJ"
        >
          <div
            className="TableCell__StyledTableCellContent-sc-110j155-1 klobR"
          >
            <span
              className="Text-yxw4o9-0 bIIgvx"
            >
              loading
            </span>
          </div>
        </td>
      </tr>
      <tr
        className="TableRow__StyledTableRow-sc-3jn6yz-0 cFtmyC"
      >
        <td
          className="TableCell__StyledTableCell-sc-110j155-0 bqABXJ"
        >
          <div
            className="TableCell__StyledTableCellContent-sc-110j155-1 klobR"
          >
            <span
              className="Text-yxw4o9-0 bIIgvx"
            >
              0x24c810eaa67a6ec5e84c75ad6656a1e1af14b6d1dcf69990079dc78e336563f7
            </span>
          </div>
        </td>
        <td
          className="TableCell__StyledTableCell-sc-110j155-0 bqABXJ"
        >
          <div
            className="TableCell__StyledTableCellContent-sc-110j155-1 klobR"
          >
            <span
              className="Text-yxw4o9-0 bIIgvx"
            >
              0x16C598eE4540b73898dEd6684e78774928cE2e88
            </span>
          </div>
        </td>
        <td
          className="TableCell__StyledTableCell-sc-110j155-0 bqABXJ"
        >
          <div
            className="TableCell__StyledTableCellContent-sc-110j155-1 klobR"
          >
            <span
              className="Text-yxw4o9-0 bIIgvx"
            >
              0x7C39a481E17CCAfbe8e50Ec1388fDBB01Bbb3bf4
            </span>
          </div>
        </td>
        <td
          className="TableCell__StyledTableCell-sc-110j155-0 bqABXJ"
        >
          <div
            className="TableCell__StyledTableCellContent-sc-110j155-1 klobR"
          >
            <span
              className="Text-yxw4o9-0 bIIgvx"
            >
              loading
            </span>
          </div>
        </td>
      </tr>
    </tbody>
  </table>,
  <br />,
]
`;
